packages/oauth/oauth-types/src/oauth-client-id-loopback.ts

```diff
@@ -15,44 +18,89 @@ export function isOAuthClientIdLoopback<C extends OAuthClientId>(
   }
 }
 
-export function parseOAuthLoopbackClientId(clientId: OAuthClientId): URL {
-  const url = parseOAuthClientIdUrl(clientId)
-
-  // Optimization: cheap checks first
+export function assertOAuthLoopbackClientId(
+  clientId: string,
+): asserts clientId is OAuthClientIdLoopback {
+  void parseOAuthLoopbackClientId(clientId)
+}
 
-  if (url.protocol !== 'http:') {
-    throw new TypeError('Loopback ClientID must use the "http:" protocol')
+// @TODO: should we turn this into a zod schema? (more coherent error with other
+// validation functions)
+export function parseOAuthLoopbackClientId(clientId: string): {
+  scope?: OAuthScope
+  redirect_uris?: [string, ...string[]]
+} {
+  if (!clientId.startsWith(OAUTH_CLIENT_ID_LOOPBACK_URL)) {
+    throw new TypeError(
+      `Loopback ClientID must start with "${OAUTH_CLIENT_ID_LOOPBACK_URL}"`,
+    )
+  } else if (clientId.includes('#', OAUTH_CLIENT_ID_LOOPBACK_URL.length)) {
+    throw new TypeError('Loopback ClientID must not contain a hash component')
   }
 
-  if (url.hostname !== 'localhost') {
-    throw new TypeError('Loopback ClientID must use the "localhost" hostname')
+  const queryStringIdx =
+    clientId.length > OAUTH_CLIENT_ID_LOOPBACK_URL.length &&
+    clientId[OAUTH_CLIENT_ID_LOOPBACK_URL.length] === '/'
+      ? OAUTH_CLIENT_ID_LOOPBACK_URL.length + 1
+      : OAUTH_CLIENT_ID_LOOPBACK_URL.length
+
+  if (clientId.length === queryStringIdx) {
+    return {} // no query string to parse
   }
 
-  if (url.hash) {
-    throw new TypeError('Loopback ClientID must not contain a fragment')
+  if (clientId[queryStringIdx] !== '?') {
+    throw new TypeError('Loopback ClientID must not contain a path component')
   }
```

Comment:
Do we handle the case that there's no path component and also no query params?

Reply:
Yep, in the previous check (`if (clientId.length === queryStringIdx)`)